scripthost/src/main/kotlin/com/commonwealthrobotics/bowlerkernel/scripthost/Session.kt

```diff
@@ -0,0 +1,271 @@
+                                    else -> throw IllegalStateException("Unhandled message type: $msg")
+                                }
+
+                                logger.debug { "Result of handling client message: $result" }
+
+                                if (msg.hasRunRequest()) {
+                                    if (result is Either.Left) {
+                                        logger.warn(result.a) { "Script did not run successfully." }
+                                        send(
+                                            sessionServerMessage {
+                                                errorBuilder.requestId = msg.runRequest.requestId
+                                                errorBuilder.description =
+                                                    "Script did not run successfully: ${result.a.localizedMessage}"
+                                            }
+                                        )
+                                    }
+                                }
+
+                                true
+                            }
+                        }
+                    }
+                }
+            } while (cont)
+        } catch (t: Throwable) {
+            logger.error(t) { "Unhandled error in session" }
+            throw t
+        } finally {
+            getKoin().unloadModules(modules)
+            logger.debug { "Session ended" }
+        }
+    }.consumeAsFlow()
+
+    init {
+        require(coroutineScope != GlobalScope) {
+            "A session may not be started in GlobalScope."
+        }
+
+        @Suppress("TooGenericExceptionCaught") // Okay because we just log and rethrow
+        try {
+            getKoin().loadModules(modules)
+        } catch (t: Throwable) {
+            logger.error(t) { "Unhandled error when constructing session" }
+            throw t
+        }
+    }
+
+    /**
+     * Handles a [RunRequest].
+     *
+     * @param runRequest The [RunRequest]
+     * @param sessionChannel A channel to send messages to the client.
+     * @return The result of loading and running the script.
+     */
+    private suspend fun runRequest(
+        runRequest: RunRequest,
+        sessionChannel: SendChannel<SessionServerMessage>
+    ): Either<Throwable, Any?> {
+        val scriptLoader = getKoin().get<ScriptLoader>()
+
+        val script = sessionChannel.withTask(
+            runRequest.requestId,
+            nextTaskID.getAndIncrement(),
+            "Initializing ${runRequest.file.path}"
+        ) {
+            val script = scriptLoader.resolveAndLoad(runRequest.file, runRequest.devsList, runRequest.environmentMap)
+            scriptRequestMap[runRequest.requestId] = script
+            script
+        }
+
+        return sessionChannel.withTask(
+            runRequest.requestId,
+            nextTaskID.getAndIncrement(),
+            "Running ${runRequest.file.path}"
+        ) {
+            script.map {
```

Comment:
This means you actually call into `withTask` even if the script loading fails (i.e. `script` is `Left`). If this is intentional, it should be documented here, because it seems unintuitive (to me at least).

Reply:
Yeah, I was being lazy and thought it was fine. I will change it.

Reply:
I couldn't write `script.map { ... withTask .. }` before arrow 0.11. Now I can.